### ISSUE TITLE
[filter-effects-2] User feedback: backdrop-filter is way less useful because it's tied to backdrop and not to background

### ISSUE THREAD

**Issue by ell1e** [w3c/fxtf-drafts#408]
Okay, so I just played around with backdrop-filter and at first it was really cool: I can have a modern milk glass effect with stuff blurred behind, and it "just works"?? How cool is that?

But then I discovered the very inconvenient weirdness of its layout effect: backdrop-filter apparently requires all child elements to be contained and therefore messes up all absolute/fixed/relative children that previously "escaped" their milk glass background parent, and contains them inside. While you might argue that this is not an issue since I can just move affected child elements out of their parent, this feels like a huge limitation for websites designed to have "clean DOM" like this:

```<!DOCTYPE HTML>
<html lang="en">
<head>...</head>
<body>
    <content>
        <img id="main-logo-small" src="/horse64logo.png" alt="Horse64 Logo"/>
        <h1>horse64, simple and versatile.</h1>
        <p><a href="/download">Try <em>Horse64</em> now</a>, a high-level language ...
```
![scr](https://user-images.githubusercontent.com/64124388/95707376-e428f480-0c48-11eb-8ef7-17dc44b961c1.png)

*(the white round corner content box with the main text inside is `<content>`, and as you can see, the `<h1>` heading is clearly placed outside of it)*

The problem here, if it isn't obvious, is that it's IMHO a very useful trick to be able to move children outside their parent in layouting to "loosen" the strict what-contains-what relationship, and therefore avoid unnecessary clutter in the DOM with tons of separate containers just for visual layouting when on a semantic level they really aren't desired to be there.

backdrop-filter breaks this, apparently to get some sort of nice contained "backdrop rectangle" with all children inside which would be ruined by letting them escape like that, so it doesn't seem to be allowed.

I don't know why exactly this decision was made, but as seen with the example above this can be extremely inconvenient. And why I don't get it as a user is, background transparency doesn't have this issue, and backdrop-filter effects similarly only seem to interact with what is *behind* the element, but *not* actually the backdrop itself much - e.g. the blur doesn't blur the element itself into its background, rather just what is *behind* it. So why does it even need to be tied to a neatly contained backdrop like that? Why can't it just let all childen be wherever, and similarly to background transparency it just applies to the clear bounds of the parent and if any children escapes then just ignore it for the effect? The way it is done seems really limiting and a bit disappointing with how it clashes with some more unusual layouts. It's just sad to need to pick between a nice semantically clean DOM, and actually having it look like I want it to.

I just wanted to give you this feedback hoping it could be changed, by e.g. moving it to a background-filter instead that is not tied to an invasive layouting change like that. Maybe that is not technically feasible, I don't know. But I sure wish it was how it was done instead.

Or is there a workaround for this that doesn't involve cluttering the DOM I'm not aware of? I really can't think of any

**Comment by mfreed7** [comment]
Thanks for the comment. If I understand your point, it is that the children ("C") of the element ("B") with `backdrop-filter` applied can be positioned outside the bounds of B, e.g. with absolute positioning. And then in that case, the backdrop behind C isn't backdrop-filtered. Is that right?

If that's the issue, this is actually mostly for convenience to the author. The alternative is to filter **absolutely everything** on the page that lives below the z-index of B, and that would often not be what the author is intending. Typical examples for backdrop filter are menu bars (that blur the content as it scrolls under) and dialog boxes (that blur the contents behind the dialog). If you're looking for a way to filter absolutely everything on the page, it is easier to do something like `<style> html {filter: blur(5px)} </style>`. I.e. just filter everything.

I don't know of a great workaround for what you're trying to accomplish, other than making the C elements siblings, instead of children, of B, and apply `backdrop-filter` to C also. Would that not work?

**Comment by ell1e** [comment]
> And then in that case, the backdrop behind C isn't backdrop-filtered. Is that right?

No. The problem is that C's positioning is completely messed up (by spec, because it limits children to the area of a backdrop'ed parent).

> I don't know of a great workaround for what you're trying to accomplish, other than making the C elements siblings, instead of children, of B, and apply backdrop-filter to C also. Would that not work?

This messes up the markup. You can see this in the code example I coded, I would be forced to move the `<h1>` outside of the `<content>` tag. This really ruins the semantic meaning, therefore I think this backdrop restriction is unreasonable.

I absolutely don't care if C is backdrop-filtered or not - in fact, I would expect it not to be as you suggested, if it is outside the area of B. But that is not the problem, the problem is that it isn't even *allowed* to be outside the area of B.

**Comment by SebastianZ** [comment]
Without a live example and reference to the specification, it took me quite some time to find out what the issue actually is. As I understand it, @ell1e refers to this sentence in the ["Filtering and Clipping" section](https://drafts.fxtf.org/filter-effects-2/#backdrop-filter-operation):

> A computed value of other than none results in the creation of both a stacking context [CSS21] and a Containing Block for absolute and fixed position descendants, unless the element it applies to is a document root element in the current browsing context.

So in https://jsfiddle.net/SebastianZ/jp5ts2mv/ the "Headline" text is positioned differently when `backdrop-filter` is applied. If it is defined, `<section>` becomes a containing block and the `<h1>` is positioned relatively to it, without it, the `<h1>`  is positioned absolutely.

To authors this is an unexpected and undesired behavior. And it's unclear why this behavior is copied over from the `filter` property. For the `filter` property it might still be clear because it applies the filters to the element itself, though `backdrop-filter` filters what's behind the element, so this shouldn't have an effect on the layout of the element's children.

Sebastian

**Comment by ell1e** [comment]
> To authors this is an unexpected and undesired behavior. And it's unclear why this behavior is copied over from the filter property. For the filter property it might still be clear because it applies the filters to the element itself, though backdrop-filter filters what's behind the element, so this shouldn't have an effect on the layout of the element's children.

Yes, exactly. And for something that just applies to the background, this has IMHO quite severe and undesirable layout limitation implications. Basically, it makes many tricks to keep the HTML markup simple and semantically meaningful impossible.

**Comment by mfreed7** [comment]
> As I understand it, @ell1e refers to this sentence in the ["Filtering and Clipping" section](https://drafts.fxtf.org/filter-effects-2/#backdrop-filter-operation):
> 
> > A computed value of other than none results in the creation of both a stacking context [CSS21] and a Containing Block for absolute and fixed position descendants, unless the element it applies to is a document root element in the current browsing context.
> 
> So in https://jsfiddle.net/SebastianZ/jp5ts2mv/ the "Headline" text is positioned differently when `backdrop-filter` is applied. If it is defined, `<section>` becomes a containing block and the `<h1>` is positioned relatively to it, without it, the `<h1>` is positioned absolutely.
> 
> To authors this is an unexpected and undesired behavior. And it's unclear why this behavior is copied over from the `filter` property. For the `filter` property it might still be clear because it applies the filters to the element itself, though `backdrop-filter` filters what's behind the element, so this shouldn't have an effect on the layout of the element's children.

Ahh, I see, the question is about the forced containing block. That is there for the same reasons as for `filter`, and it is more fundamental than just implementation convenience. @chrishtr can probably explain more succinctly than I can.
